### PR TITLE
Added AnimePahe

### PR DESCRIPTION
**Pull request by ArjixWasTaken**

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
It works as is.
Should I also add "Random Anime" to the homepage?
I can pick like 15 random anime, and show them in a list.

**Comment by ArjixWasTaken** [comment]
I guess the kwik sorcery could be moved to a separate extractor.
But I had a bad experience making one and I don't want to touch that code.

If you want to you can convert the kwik code to an extractor.
Although I can't think of any site other than AnimePahe that uses kwik....

**Comment by ArjixWasTaken** [comment]
And uh....how do you do that homepage thingy with the huge slideshow of movies/shows in trailers.to and wcostream?

**Review by LagradOst — COMMENTED** [review]
Not provider sufix?

**Comment by ArjixWasTaken** [comment]
> Not provider sufix?

uh

**Comment by ArjixWasTaken** [comment]
You saw nothing

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+                val response = khttp.get("https://animepahe.com/")
+                cookies = response.cookies
+                true
+            } catch (e: Exception) {
+                false
+            }
+        }
+        val YTSM = "ysmm = '([^']+)".toRegex()
+
+        val KWIK_PARAMS_RE = """\(\"(\w+)\",\d+,\"(\w+)\",(\d+),(\d+),\d+\)""".toRegex()
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:33]: Not `Regex("""\(\"(\w+)\",\d+,\"(\w+)\",(\d+),(\d+),\d+\)""")` :pensive:

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+            val doc = Jsoup.parse(html)
+
+            val title = doc.selectFirst(".anime-poster img")?.attr("alt")?.replace("Poster of ", "")
+            val japTitle = doc.selectFirst("h2.japanese")?.text()
+            val poster = doc.selectFirst(".anime-poster a").attr("href")
+
+            val TvType = doc.selectFirst("""a[href*="/anime/type/"]""")?.text()
+
+            val trailer: String? = if (html.contains("https://www.youtube.com/watch")) {
+                YOUTUBE_VIDEO_LINK.find(html)?.destructured?.component1()
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:239]: pretty sure you can just do 
```kotlin
val (trailer): String? = if (html.contains("https://www.youtube.com/watch")) {
                YOUTUBE_VIDEO_LINK.find(html)?.destructured
```

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+                } else { null },
+                ArrayList(),
+                malId,
+                anilistId,
+                null,
+                trailer
+            )
+        } catch (e: Exception) {
+            println("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
+            e.printStackTrace()
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:289]: Use normalSafeApiCall

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+            throw e
+        }
+    }
+
+
+    private fun isNumber(s: String?): Boolean {
+        return s?.toIntOrNull() != null
+    }
+
+    private fun pow(base: Int, expo: Int): Long{
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:299]: try with build in pow and see if it works :pray:

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+    private fun bypassAdfly(adflyUri: String): String {
+        if (!generateSession()) {
+            return bypassAdfly(adflyUri)
+        }
+
+        var responseCode = 302
+        var adflyContent: khttp.responses.Response? = null
+
+        while (responseCode != 200) {
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:436]: not keen on risking an infinite request loop, add a max if you haven't and i missed it

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+
+        var code = 419
+
+        while (code != 302) {
+            content = khttp.post(
+                uri,
+                allowRedirects = false,
+                data=mapOf("_token" to tok),
+                headers=mapOf("referer" to fContent.url),
+                cookies=cookieStrToMap(fContent.headers.getValue("set-cookie").replace("path=/,", ""))
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:462]: can't you do `fContent = khttp.get(bypassAdf..... ).cookies` and then pass it in a second khttp request without anything else?

**Review by Blatzar — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 1 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:33]: virtually no difference but ok

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 2 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:239]: shh

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 3 above)
Comment (LagradOst) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:289]: > 
> 
> Use normalSafeApiCall

but no, that should be handled for you, you should not use try catch in the providers if you dont want to display a specified error message

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 2 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:239]: uhh, remove the if statement then

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 3 above)
Comment (Blatzar) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:289]: Need to steal that safety for Shiro

Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:289]: kk

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 5 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:436]: I'll make it 20 then.
It rarely goes above 2 iterations.

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 6 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:462]: uhh, no

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt (hunk 2 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt:239]: fucking fails

**Comment by ArjixWasTaken** [comment]
Ok fixed a bug with the episode link generation.
(kotlin int ranges are weird, `0..10` will be 11 iterations...)

**Pull request merged**